Create instance, physical device and device from requirements

sheaf opened the pull request:
Adds some code to the  `vulkan-utils` package to create instances, physical devices and devices from `Requirements` rather than a bunch of names. Should support all the functionality it did before, as this patch also adds support for optional features in the `Requirements` module of the `vulkan` package.    

I changed the approach for creating an instance with debugging enabled, by providing a function that sets the same debugging requirements as were used before. Then those requirements can be passed to the function which creates the instance.

expipiplus1 commented on utils/src/Vulkan/Utils/Initialization.hs:
```diff
@@ -1,146 +1,235 @@
+          ( "Optional preference: minimum Vulkan API version " <> showVersion optVersion <> ",\n\
+            \but the version specified by the application is " <> showVersion givenVersion <> "."
+          )
+        )
```
Yeah, we should use some more structured error type here, can do that later on though.

expipiplus1 commented on utils/src/Vulkan/Utils/Initialization.hs:
```diff
@@ -1,146 +1,235 @@
+--   - returns 'Nothing'.
+--
+-- __Note__: This function assumes that the context (i.e. the Vulkan instance) already meets the 'Requirements',
+-- in particular that the Vulkan API version is set to at least 1.1 if the 'Requirements' impose this restriction.
```
could use `enumerateInstanceVersion`?

sheaf replied:
Ok.

sheaf replied:
Done.

expipiplus1 commented on utils/src/Vulkan/Utils/Initialization.hs:
```diff
@@ -182,43 +271,89 @@ physicalDeviceName =
+    , enableFeaturesOtherThan = enableFeaturesOtherThan :: HashSet ByteString -> PhysicalDeviceFeatures2 feats
+    }
+  )
+  don'tEnableThese
```
nice

expipiplus1 commented on utils/vulkan-utils.cabal:
```diff
@@ -62,6 +63,7 @@ library
     , text
     , transformers
     , typed-process
+    , unordered-containers
```
Could you edit the `package.yaml`s instead and run `hpack`? (I already did that here: https://github.com/expipiplus1/vulkan/commit/28c67d900ae0ad8de71ca2f716fc4beff35660fb)

expipiplus1 commented:
I think it would be really good to replicate what the queue assignment
stuff does. It takes the requirements packed into any `Traversable` and returns
to the user the same shape but with proof that the queues have been assigned
correctly at each index (in the form of a program to get that queue from the
device, and the device params).

The way that would work here would be along the lines of:

```haskell
createInstanceFromRequirements
  :: Traversable f
  => f Requirements -- optional requirements
  -> InstanceCreateInfo es
  -> io (f RequirementResult, Maybe Instance)

data RequirementResult
  = Satisfied
  | Unsatisfied Reason
  | UnsatisfiedOptional Reason
```

To be used like

```haskell
data MyReqs a = MyReqs
  { hasRaytracing  :: a -- Required
  , hasDebugPrintf :: a -- Optional
  }

myReqs :: MyReqs Requirements
myReqs = MyReqs (requireBlah rtRequirements)
                (optionalBlah debugReqs)

main = do
  (rs, mbInst) <- createInstanceFromRequirements myReqs zero
  inst <- case mbInst of
    Just inst -> pure inst
    Nothing -> do
      sayErr "Couldn't create an Instance"
      traverse (sayErr . prettyReason) rs
  if hasDebugPrintf rs == Satisfied
    then doMyDebugProgram
    else theOtherCodePath
```

Also, isn't an optional requirement a bit of an oxymoron?

I realise that the types don't fit 100% above but you get the gist I hope

We might even want some separate 'proof' instead of `Satisfied`, for example a
function to add the features to the createinfo chain, and then split the
creation and finding requirements function.

```haskell
data RequirementResult
  = Satisfied (SomeStruct InstanceCreateInfo -> SomeStruct InstanceCreateInfo) -- or something
  ...

main = do
  rs <- getInstanceCreateInfo myReqs
  -- insert check error here
  let i :: InstanceCreateInfo '[]
      createInfo = appEndo i $ foldMap Endo [f | Satisfied f <- toList rs]
  withInstance createInfo
  ...
```

expipiplus1 commented:
Oh, and having the optionalness inside the requirement seems a bit odd, what do you think about having it on the outside, i.e. requirements are requirements, however where they are used, two sets of them are passed: one optional set of Requirements and one non-optional.

(This does complicate layer ordering however)

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -36,34 +42,60 @@ import GHC.Exts
+{-# COMPLETE Optionally, Necessarily #-}
+pattern Optionally, Necessarily :: Requirement -> ( Requirement, Bool )
+pattern Optionally  r = ( r, True  )
+pattern Necessarily r = ( r, False )
```
No, let me know what to name it (possibly changing the name of `Requirement`) and I'll make the change.

sheaf replied:
Done.

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -184,162 +237,246 @@ data Requirements where
+       --
+       -- Argument: optional features that should __not__ be enabled.
+       -- These must be the missing optional features returned by 'missingFeatures'.
+       , enableFeaturesOtherThan :: HashSet ByteString -> PhysicalDeviceFeatures2 feats
```
`missingFeatures` returns the features which can't be enabled on the device. Then `enableFeaturesOtherThan` should be passed the optional features that can't be enabled. I wrote an explanation to this effect in the comments to `createDeviceFromRequirements`.    

This way made the most sense to me, as if we want this to be done internally by the function which manages requirements, we would have to change `enableFeaturesOtherThan` to `enableFeatures :: PhysicalDeviceFeatures2 feats -> PhysicalDeviceFeatures2 feats``` which takes the queried `PhysicalDeviceFeatures2` and returns what amounts to `zero` with the the features enabled. So, if we want to keep choosing the physical device and creation of the logical device separate, the user would still have to end up passing an argument from one to the other as they do now.

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -184,162 +237,246 @@ data Requirements where
-       , missingFeatures    :: PhysicalDeviceFeatures2   feats -> HashSet ByteString
+       --
+       -- Argument: optional settings that the function will __not__ attempt to enable.
+       , enableSettingsOtherThan :: HashSet ByteString -> Chain setts
```
See explanation for `enableFeaturesOtherThan`. `vulkan-utils` isn't currently using this, but I think the same logic should apply, perhaps this time involving the optional layer names that were not able to be enabled.

sheaf replied:
I removed this, as you explained to me that it is always safe to enable these.

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -184,162 +237,246 @@ data Requirements where
+        (const ())
+        (const $ ReqAndOpt HashSet.empty HashSet.empty)
+        (const zero)
+        (const $ ReqAndOpt HashSet.empty HashSet.empty)
```
It would require proving that any two structure chains can be combined into a single one, which sounds like it would be quite difficult. It can be made into an instance of `zero` using this as `zero`, obviously.

sheaf commented on utils/src/Vulkan/Utils/Initialization.hs:
```diff
@@ -1,146 +1,235 @@
-      (EXT_DEBUG_UTILS_EXTENSION_NAME : requiredExtensions)
-      (EXT_VALIDATION_FEATURES_EXTENSION_NAME : optionalExtensions)
-      instanceCreateInfo'
-    _ <- withDebugUtilsMessengerEXT inst
```
As mentioned in the PR description: see `addDebugRequirements`.

sheaf replied:
Addressed this.

sheaf commented on utils/src/Vulkan/Utils/Initialization.hs:
```diff
@@ -1,146 +1,235 @@
+  :: forall m
+   . MonadResource m
+  => Requirements
+  -> InstanceCreateInfo '[]
```
Yes.

sheaf commented on utils/src/Vulkan/Utils/Initialization.hs:
```diff
@@ -1,146 +1,235 @@
+   . MonadResource m
+  => Requirements
+  -> InstanceCreateInfo '[]
+  -> m (Instance, [ByteString])
```
Ok.

sheaf replied:
Marking as done as you mentioned you would try to tackle this.

sheaf commented:
> Also, isn't an optional requirement a bit of an oxymoron?
Yes, that was because I added the "optional" functionality in a bit of an ad-hoc way, please suggest a renaming for `Requirement`, `Optional Requirement`, and `Optionally/Necessarily`.

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -94,7 +126,9 @@ data Requirement where
-    => { setInstanceSetting :: struct -> struct }
+    => { settingName        :: ByteString
+       , setInstanceSetting :: struct -> struct
+       }
```
> I think it would be really good to replicate what the queue assignment
> stuff does. It takes the requirements packed into any `Traversable` and returns
> to the user the same shape but with proof that the queues have been assigned
> correctly at each index (in the form of a program to get that queue from the
> device, and the device params).
> 
> The way that would work here would be along the lines of:
> 
> ```haskell
> createInstanceFromRequirements
>   :: Traversable f
>   => f Requirements -- optional requirements
>   -> InstanceCreateInfo es
>   -> io (f RequirementResult, Maybe Instance)
> 

Ok, but the point was that the type `es` is created by inspecting the `Requirements`, so the user doesn't have to come up with the structure chain. I was hoping to entirely avoid having the user need to come up with structure chains if possible.

sheaf replied:
I changed it so you directly specify the struct instead of providing a modification function, and I added a comment. I think that should be enough.

sheaf commented:
I think it would be really good to replicate what the queue assignment
stuff does. It takes the requirements packed into any Traversable and returns
to the user the same shape but with proof that the queues have been assigned
correctly at each index (in the form of a program to get that queue from the
device, and the device params).

The way that would work here would be along the lines of:

createInstanceFromRequirements
  :: Traversable f
  => f Requirements -- optional requirements
  -> InstanceCreateInfo es
  -> io (f RequirementResult, Maybe Instance)

data RequirementResult
  = Satisfied
  | Unsatisfied Reason
  | UnsatisfiedOptional Reason

sheaf commented:
I don't think ` f Requirement -> ... -> io ( f RequirementResult, ... )` would be possible with the way the various usages of `Requirements` are split up (because we want to collect up the requirements a single time). I think instead we could simply change `partitionOptReqIO` to return `RequirementResult` instead of throwing errors/warnings?

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -184,162 +237,246 @@ data Requirements where
+       -- | Required and optional minimum Vulkan API version.
+         version                 :: ReqAndOpt Word32
+       -- | Required and optional Vulkan layer names and their minimum versions.
+       , instanceLayers          :: ReqAndOpt (HashMap ByteString Word32)
```
OK, I didn't know that. Are there some layers that need to be enabled before others then?

expipiplus1 replied:
not necessarily, but consider where `api_dump` goes for example, you might want it first in the list (next to the application) to avoid capturing all the vulkan commands generated by subsequent layers.

sheaf replied:
Makes sense yeah.

sheaf replied:
Done.

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -184,162 +237,246 @@ data Requirements where
+            mbLayName
+            (ReqAndOpt HashMap.empty (HashMap.singleton extName extVer))
+            exts
+        }
```
I'll think about it.

sheaf replied:
The problem is that we have a dependent type, e.g.

```haskell
props' = case has @props @struct of { Just _ -> props; _ -> struct : props }
```
So I end up duplicating all of the code that depends on `props'` until I get to the constructor which wraps up the existential so that the type becomes non-dependent again.

sheaf replied:
I think I'll leave it as is, otherwise we have to introduce a dependent sum type that captures exactly the constraints we need, seems like it'll just make the code more confusing.

expipiplus1 commented:
> OK, but the point was that the type `es` was creating by inspecting the requirements, so the user doesn't have to manually come up with a structure chain. I was hoping that in doing so, the user would no longer need to come up with the structure chains themselves.

Ah, I just put that in there to mean any additional things the user wants in there which are easier to express extra the Requirements system, these 'es' would just be appended.

Happy to have `'[]` instead, no biggie

sheaf commented on src-manual/Vulkan/Requirements.hs:
```diff
@@ -184,162 +237,246 @@ data Requirements where
+addRequirement (RequireFeature {..}, isOptional) =
+  requireFeature isOptional featureName checkFeature enableFeature
+addRequirement (RequireProperty {..}, isOptional) =
+  requireProperty isOptional propertyName checkProperty
```
Done.

sheaf commented:
- Updated `Requirements` to separate instance requirements from device requirements.
- Updated `generate-new` to accomodate for these changes. I also made some additional changes: on my end the spaces at the start of the lines in source code quasi-quotations were messing up the generated source code (by including these extra spaces). So I removed those.
- Updated `vulkan-utils` to track those changes.

I think we're finally in a pretty decent state now, I'll pass the torch over to you.

expipiplus1 commented on generate-new/src/Bespoke.hs:
```diff
@@ -756,14 +756,14 @@ boolConversion = genRe "Bool conversion" $ do
-    boolToBool32 :: Bool -> {bool32}
-    boolToBool32 = bool {false} {true}
+boolToBool32 :: Bool -> {bool32}
+boolToBool32 = bool {false} {true}
```
lol, what happened to all my quasiquotes

sheaf replied:
On my end, the spaces at the start of the lines in the quasi quotes were being included in the generated source, which caused parse errors. No idea why the behaviour is different on my end than yours... obviously feel free to ditch this if this causes issues on your end now.

expipiplus1 replied:
weird!

expipiplus1 replied:
Perhaps this doesn't work on windows? https://github.com/expipiplus1/vulkan/blob/43222eecf54c41546bbcda1e4b467c7ba35d18a6/generate-new/src/Text/InterpolatedString/Perl6/Unindented.hs

expipiplus1 commented on generate-new/src/Render/Element/Write.hs:
```diff
@@ -283,8 +283,17 @@ renderModule out boot getDoc findModule findLocalModule (Segment modName unsorte
         let t = layoutDoc (d <> line <> line)
         if getAll (reCanFormat e)
           then liftIO (parsePrintModule brittanyConfig t) >>= \case
-            Left  _ -> error "Fail"
```
Sorry, I completely forgot I left this in!

expipiplus1 commented on generate-new/src/Render/Element/Write.hs:
```diff
@@ -283,8 +283,17 @@ renderModule out boot getDoc findModule findLocalModule (Segment modName unsorte
+            Left  err -> error $ "Fail: Brittany failed to handle module:\n" <>
+                                    T.intercalate "\n" (fmap (T.pack . showBError) err)
+                                    <> "\n\n\n" <> t
+              where showBError = \case
```
Yeah, I bumped in to this the other day, crazy that it doesn't have a show instance

sheaf replied:
Yeah, what???

expipiplus1 commented:
I pushed a branch spirv-sheaf which is rebased on master without the QQ stuff

expipiplus1 commented:
Merged as part of the raytracing branch.

expipiplus1 commented:
Will release as 3.8 soon
